crates/ln-dlc-node/src/ldk_node_wallet.rs

```diff
@@ -180,58 +188,121 @@ where
         Ok(utxos)
     }
 
-    pub fn get_utxos_for_amount(
+    pub fn get_utxos_for_dlc_funding_transaction(
         &self,
         amount: u64,
-        lock_utxos: bool,
-        network: Network,
+        fee_rate: Option<u64>,
+        should_lock_utxos: bool,
     ) -> Result<Vec<Utxo>> {
-        let utxos = self.get_utxos()?;
-        // get temporarily reserved utxo from in-memory storage
+        let network = {
+            let bdk = self.bdk_lock();
+            bdk.network()
+        };
+
+        let fee_rate = fee_rate.map(|fee_rate| fee_rate as f32).unwrap_or_else(|| {
+            self.get_fee_rate(ConfirmationTarget::Normal)
+                .as_sat_per_vb()
+        });
+
+        // Get temporarily reserved UTXOs from in-memory storage.
         let mut reserved_outpoints = self.locked_outpoints.lock();
 
-        // filter reserved utxos from all known utxos to not accidentally double spend and those who
-        // have actually been spent already
+        let utxos = self.get_utxos()?;
```

Comment:
🤔 Could we run into race conditions here? e.g. if two dlc channels are setup at the same time and they pick the same utxos? i guess we need some locking around selecting and reserving utxos.

Reply:
Hmm, there is a mutex around locked_outpoints.

Reply:
Actually, I think you have a point, but for a different scenario.

Two independent calls to `get_utxos_for_dlc_funding_transaction` cannot run into a race condition, because the `locked_outpoints` mutex will be held by one of them first, and it is only dropped at the end of the function.

But a different method not protected by this specific mutex could pick the same UTXOs and try to use them too. I think Philipp already fixed this problem with d9178a45354f271fc29fad917cdabd35fc8048ec, but it wouldn't be hard to misuse this mutex.

Maybe with the new version of `bdk` UTXO management is a bit easier.

Reply:
Right, I missed the lock just a few lines above 🙈